src/operation/insert/mod.rs

```diff
@@ -65,25 +67,37 @@ impl<T: Serialize> Operation for Insert<T> {
             .take(description.max_write_batch_size as usize)
             .enumerate()
         {
-            let mut doc = bson::to_document(d)?;
-            let id = doc
-                .entry("_id".to_string())
-                .or_insert_with(|| {
-                    self.inserted_ids
-                        .get(i)
-                        .cloned()
-                        .unwrap_or_else(|| Bson::ObjectId(ObjectId::new()))
-                })
-                .clone();
-
-            let doc_size = bson_util::array_entry_size_bytes(i, &doc);
+            let mut doc = bson::to_vec(d)?;
+            let id = match bson_util::raw_get(doc.as_slice(), "_id")? {
+                Some(b) => b,
+                None => {
+                    let oid = ObjectId::new();
+
```

Comment:
If it's not in there, we write the new element into the beginning of the document. Once we have the raw document API we could update this to use it.

Reply:
Can we add a TODO comment for this?

Reply:
done